firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java

```diff
@@ -103,13 +131,105 @@ public BackfillScheduler newScheduler(AsyncQueue asyncQueue, LocalStore localSto
     return new BackfillScheduler(asyncQueue, localStore);
   }
 
-  // TODO(indexing): Figure out which index entries to backfill.
-  public Results backfill() {
-    int numIndexesWritten = 0;
+  public Results backfill(LocalStore localStore) {
+    // TODO(indexing): Handle field indexes that are removed by the user.
+    fetchNewFieldIndexes();
+
+    int numIndexesWritten = writeIndexEntries(localStore);
     int numIndexesRemoved = 0;
     return new Results(/* hasRun= */ true, numIndexesWritten, numIndexesRemoved);
   }
 
+  private void fetchNewFieldIndexes() {
+    List<FieldIndex> fieldIndexes = indexManager.getFieldIndexes(lastFieldIndexId + 1);
+
+    if (fieldIndexes.size() > 0) {
+      fieldIndexQueue.addAll(fieldIndexes);
+
+      // Since index configuration entries are sorted by index id, we can set the last fetched id to
+      // the id of the last field index in the array.
+      lastFieldIndexId = fieldIndexes.get(fieldIndexes.size() - 1).getIndexId();
+    }
+  }
+
+  /** Writes index entries based on the FieldIndexQueue. Returns the number of entries written. */
+  private int writeIndexEntries(LocalStore localStore) {
+    int numIndexesWritten = 0;
+    SnapshotVersion lastReadTime = null;
+
+    // TODO(indexing): Track the number of entries written and process more field indexes if
+    // we're under the maximum.
+    int lastArrayIndex = Math.min(maxFieldIndexesToProcess, fieldIndexQueue.size());
+    List<FieldIndex> fieldIndexesToProcess = fieldIndexQueue.subList(0, lastArrayIndex);
+
+    // Create a map of each collection group to all the FieldIndexes on it.
+    Map<String, ArrayList<FieldIndex>> collectionToFieldIndexes = new HashMap<>();
+
+    // Map of field index to the most recent snapshot version written.
+    Map<FieldIndex, SnapshotVersion> fieldIndexToNewestVersion = new HashMap<>();
+
+    for (FieldIndex fieldIndex : fieldIndexesToProcess) {
+
+      // Use the lowest version number as the base.
+      lastReadTime =
+          lastReadTime != null
+              ? fieldIndex.getVersion().compareTo(lastReadTime) < 0
+                  ? fieldIndex.getVersion()
+                  : lastReadTime
+              : fieldIndex.getVersion();
```

Comment:
This can likely be simplified if you directly update the FieldIndex version.

Reply:
I realize that since we're processing the collection groups in order, we can move the logic down to fetch the lowest version for all fieldindexes mapped to a collection group. 

For example, if we have multiple field indexes that map to a collection group, we can load documents starting from the lowest snapshot version present on the field indexes.